### PR TITLE
Implement RegularExpressionPatternRule

### PR DESCRIPTION
**Pull request by enumag**
Solves #108

### CODE REVIEW

File: src/Rule/Nette/RegularExpressionPatternRule.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		$patternNode = $staticCall->getArgs()[1]->value;
+		$patternType = $scope->getType($patternNode);
+
+		$patternStrings = [];
+
+		foreach (TypeUtils::getConstantStrings($patternType) as $constantStringType) {
+			$patternStrings[] = $constantStringType->getValue();
+		}
+
+		foreach ($patternType->getConstantArrays() as $constantArrayType) {
```
Comment (enumag) [inline src/Rule/Nette/RegularExpressionPatternRule.php:81]: I copied this from the rule in PHPStan but I'm wondering if `TypeUtils::getConstantArrays()` should be used here instead similar to `TypeUtils::getConstantStrings()` a few lines above. What's the difference?

Comment (enumag) [reply src/Rule/Nette/RegularExpressionPatternRule.php:81]: Failed tests say that `$patternType->getConstantArrays()` doesn't exist in older version. Should I increase the required PHPstan version to 1.9.0 where it was added or should I use `TypeUtils`?

Comment (enumag) [reply src/Rule/Nette/RegularExpressionPatternRule.php:81]: Ah just noticed that `TypeUtils::getConstantArrays()` is actually deprecated so I'll increase the required version then.

**Comment by enumag** [comment]
@ondrejmirtes What about this one? I'd really like to have the regular expressions validated.

**Comment by ondrejmirtes** [comment]
@enumag I have 140 unread emails in my inbox. There are other people waiting too + I have my own priorities what to do with the projects. Sorry.

**Comment by enumag** [comment]
Understood, sorry.

File: src/Rule/Nette/RegularExpressionPatternRule.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+	/**
+	 * @return string[]
+	 */
+	private function extractPatterns(StaticCall $staticCall, Scope $scope): array
+	{
+		if (!$staticCall->class instanceof Node\Name || !$staticCall->name instanceof Node\Identifier) {
+			return [];
+		}
+		$className = (string) $staticCall->class;
```
Comment (ondrejmirtes) [inline src/Rule/Nette/RegularExpressionPatternRule.php]: Call `$scope->resolveName()`. This makes for example `self` resolved into a real class name.

File: rules.neon [hunk 3]
```diff
@@ -16,6 +16,7 @@ parametersSchema:
 
 rules:
 	- PHPStan\Rule\Nette\DoNotExtendNetteObjectRule
+	- PHPStan\Rule\Nette\RegularExpressionPatternRule
```
Comment (ondrejmirtes) [inline rules.neon]: Needs to be added only in bleedingEdge.

File: src/Rule/Nette/RegularExpressionPatternRule.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+		if (!isset($staticCall->getArgs()[1])) {
+			return [];
+		}
+		$patternNode = $staticCall->getArgs()[1]->value;
+		$patternType = $scope->getType($patternNode);
+
+		$patternStrings = [];
+
+		foreach (TypeUtils::getConstantStrings($patternType) as $constantStringType) {
```
Comment (ondrejmirtes) [inline src/Rule/Nette/RegularExpressionPatternRule.php]: Use `Type::getConstantStrings()` instead.

Comment (ondrejmirtes) [reply src/Rule/Nette/RegularExpressionPatternRule.php]: Require `^1.9.4` for that.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: rules.neon (hunk 3 above)
Comment (enumag) [reply rules.neon]: :thinking: How can I do that?

Comment (enumag) [reply rules.neon]: I tried to copy something from phpstan-doctrine. Hopefully it's correct.

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**